secrets.go

```diff
@@ -0,0 +1,153 @@
+package pipeline
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/buildkite/go-pipeline/ordered"
+	"gopkg.in/yaml.v3"
+)
+
+var _ interface {
+	json.Unmarshaler
+	ordered.Unmarshaler
+} = (*Secrets)(nil)
+
+// Secrets is a sequence of secrets. It is useful for unmarshaling.
+type Secrets []*Secret
+
+// UnmarshalOrdered unmarshals Secrets from []any (sequence of secret names).
+func (s *Secrets) UnmarshalOrdered(o any) error {
+	switch o := o.(type) {
+	case nil:
+		// `secrets: null` is invalid - should be omitted entirely or use valid formats
+		return fmt.Errorf("unmarshaling secrets: secrets cannot be null")
+
+	case []any:
+		for _, c := range o {
+			switch ct := c.(type) {
+			case string:
+				secret := &Secret{
+					Key:                 ct,
+					EnvironmentVariable: &ct, // Default EnvironmentVariable to key value for simple string format
+				}
+				*s = append(*s, secret)
+
+			case map[string]any:
+				secret := &Secret{}
+
+				if key, ok := ct["key"].(string); ok {
+					secret.Key = key
+				}
+
+				if envVar, ok := ct["environment_variable"].(string); ok {
+					secret.EnvironmentVariable = &envVar
+				}
+
+				// Validate that we have at least a key
+				if secret.Key == "" {
+					return fmt.Errorf("unmarshaling secrets: secret object missing required 'key' field")
+				}
+
+				*s = append(*s, secret)
+
+			case *ordered.Map[string, interface{}]:
+				// Backend sends ordered.Map format
+				secret := &Secret{}
+
+				if keyVal, _ := ct.Get("key"); keyVal != nil {
+					if key, ok := keyVal.(string); ok {
+						secret.Key = key
+					}
+				}
```

Comment:
This logic might benefit from inverting. Given that `key` has to be present and must be a string, we could return early. Something like:

```suggestion
				keyVal, _ := ct.Get("key")
				key, ok := keyVal.(string)
				if !ok {
					return fmt.Errorf("unmarshaling secret: key must be a string, but was %T", keyVal)
				}
				secret.Key = key
```

And `key` must be non-empty, right? Which means the "default zero value"  property of various expressions in Go can be used to check for missing, empty, and incorrect type:

 ```suggestion
 				keyVal, _ := ct.Get("key")
 				// Still need two values on the left side to avoid panic
				key, _ := keyVal.(string)
				if key == "" {
					return fmt.Errorf("unmarshaling secret: key must be a non-empty string, but was %[1]T %[1]q", keyVal)
				}
				secret.Key = key
```

which would subsume the check below for `secret.Key == ""`.

Reply:
ohhh, TIL `%[1]<printf directive>` for only having to pass one value to printf 👀